### PR TITLE
WIP: Allow extending the expiry of a reservation

### LINKED ISSUE #13: Allow reservation to be extended
Currently the reservations are always deleted after 1 week.
For the next update we want to allow people to extend their reservation so they can keep their workload live for more then a week

### PR DESCRIPTION
**Pull request by serboctor**
Resolves https://github.com/threefoldtech/grid_broker/issues/13
relevant PR https://github.com/threefoldtech/jumpscaleX/pull/649

### CODE REVIEW

File: templates/grid_broker/grid_broker.py [hunk 1]
```diff
@@ -62,11 +63,32 @@ def _watch_transactions(self):
...
-                self.logger.info("transaction processed %s", tx.id)
-                # insert connection info into mail
-                if connection_info:
-                    self._send_connection_info(data['email'], connection_info)
+                if data["type"] == "extension":
+                    action = "extend"
+                    action_type = "extension"
+                    title = "Extending reservation failed"
+
+                    expiry_date = self._extend_reservation(tx, data)
```
Comment (zaibon) [inline templates/grid_broker/grid_broker.py]: why does `_extend_reservation` not return the datetime object directly instead of doing the conversion in the next line ?

File: templates/grid_broker/grid_broker.py [hunk 2]
```diff
@@ -347,19 +377,30 @@ def _is_locked(self, tx):
...
         </ul>
     </div>
 </body>
 </html>
 """
+
+_extend_template = """
+<html>
+<body>
+    <h1>Your reservation {tx_id} has been extending successfully</h1>
```
Comment (zaibon) [inline templates/grid_broker/grid_broker.py]: can we also specify what kind of reservation has been extended ?
Only the transaction ID is not really explicit.

File: templates/reservation/reservation.py [hunk 3]
```diff
@@ -30,9 +32,22 @@ def __init__(self, name, guid=None, data=None):
     def validate(self):
         if not self.data.get('creationTimestamp'):
             self.data['creationTimestamp'] = time.time()
+        if not self.data.get('expiryTimestamp'):
+            self.data['expiryTimestamp'] = time.time()
```
Comment (zaibon) [inline templates/reservation/reservation.py]: does that means all previous transaction will be marked as expired after we apply this patch ?

File: templates/reservation/reservation.py [hunk 4]
```diff
@@ -30,9 +32,22 @@ def __init__(self, name, guid=None, data=None):
...
+
+    def extend(self, duration):
```
Comment (zaibon) [inline templates/reservation/reservation.py]: This is a bit hard to review like that. Can you add a bunch of test to validate the calculation of the extended time is correct ?

File: templates/reservation/reservation.py [hunk 5]
```diff
@@ -30,9 +32,22 @@ def __init__(self, name, guid=None, data=None):
...
+        month = month % 12 + 1
+        day = min(expiry.day, calendar.monthrange(year,month)[1])
+        extended = datetime(year, month, day, expiry.hour, expiry.minute, expiry.second, expiry.microsecond)
+        self.data["expiryTimestamp"] = int(time.mktime(extended.timetuple()))
+
+        return self.data["expiryTimestamp"]
 
     @retry((Exception), tries=4, delay=5, backoff=2, logger=None)
-    def install(self):
+    def install(self,duration):
```
Comment (zaibon) [inline templates/reservation/reservation.py]: This is very unusuall to have a argument with the `install` action, why is this required ?

File: templates/reservation/reservation.py [hunk 6]
```diff
@@ -50,8 +65,9 @@ def install(self):
             raise ValueError("unsupported reservation type %s size %s" % (self.data['type'], self.data['size']))
 
         install_result = install(self.data['size'])
+        self.extend(duration)
         self.state.set('actions', 'install', 'ok')
-        return install_result
+        return install_result, self.data["expiryTimestamp"]
```
Comment (zaibon) [inline templates/reservation/reservation.py]: I don't think you can return 2 values from an action.
I think this will create an undefined behavior on the client side that doesn't not how to deal with this value.

**Review by zaibon — CHANGES_REQUESTED** [review]

File: templates/reservation/reservation.py (hunk 5 above)
Comment (serboctor) [reply templates/reservation/reservation.py]: Yea, actually I can just compute it in the grid_broker and set it in the data.

File: templates/reservation/reservation.py (hunk 3 above)
Comment (serboctor) [reply templates/reservation/reservation.py]: well, in case of install, I am depending on install setting the extending it. However, I dont like the fact that we don't set the creationtimestamp in the data before creating the service to begin with. I'll add both to the data of the service before creation in the grid_broker. If I remove it those two lines from the validate, should I then raise an error if someone tries to extend a service where the expirytimestamp isnt set? this should theoritically not happen unless someone ceates the service themselves and not through grid_broker.

File: templates/reservation/reservation.py (hunk 4 above)
Comment (serboctor) [reply templates/reservation/reservation.py]: I moved the functionality under `j.tools.time` and added a test there.
https://github.com/threefoldtech/jumpscale_lib/commit/12ec507e86ab2d0e8593ac64c5a8004d0b018642

File: templates/reservation/reservation.py (hunk 3 above)
Comment (serboctor) [reply templates/reservation/reservation.py]: Ok since I now set it in the data before creating a service, I changed the validate to raise an error if either aren't set. I think this makes more sense.

File: templates/reservation/reservation.py [hunk 7]
```diff
@@ -28,11 +27,23 @@ def __init__(self, name, guid=None, data=None):
         self.recurring_action(self._cleanup, 3600*12)  # 12h
 
     def validate(self):
-        if not self.data.get('creationTimestamp'):
-            self.data['creationTimestamp'] = time.time()
+        for key in ['creationTimestamp', 'expiryTimestamp']:
```
Comment (zaibon) [inline templates/reservation/reservation.py:48]: What about already existing services?
they won't have a value on the expiryTimestamp

Comment (serboctor) [reply templates/reservation/reservation.py:48]: May be we can have some sort of migration script/function? and for existing active reservations that have a creationtimestamp and no expiry, it sets the expiry to 1 month after the creation? Because I believe the validation needs to be there.

File: templates/grid_broker/grid_broker.py [hunk 8]
```diff
@@ -45,53 +48,99 @@ def _watch_transactions(self):
                 continue
             # try to parse the transaction data
             try:
+                malformed = True
                 data = self._parse_tx_data(tx)
+                malformed = False
                 # refund if there is not data
                 if not data:
+                    self.logger.info("no data found in transaction %s", tx.id)
+                    raise ValueError("transaction has no data")
```
Comment (zaibon) [inline templates/grid_broker/grid_broker.py]: This is not right. Here you raise cause there is no data.
So you end up in the except block line 58. And in this block you try to refund. But refund will always fail if there is not data cause you don't know the email address of the sender.

Comment (zaibon) [reply templates/grid_broker/grid_broker.py]: Also I don't really understand why you need the `malformed` variable. The only thing that happens in this try block is the `_parse_tx_data`.  So the only case where you can end up in the except block is if the parsing of the transaction failed.

Comment (zaibon) [reply templates/grid_broker/grid_broker.py]: I think we need to split this more, something like:
```python
data = None
try:
   data = parse_tx(data)
except:
   log.error

if not data:
    log error
     return

try:
    deploy
except:
    refund
```

File: templates/reservation/reservation.py (hunk 7 above)
Comment (zaibon) [reply templates/reservation/reservation.py:48]: > for existing active reservations that have a creationtimestamp and no expiry, it sets the expiry to 1 month after the creation? 

Good idea. So the template can update the data of the services automatically

File: templates/reservation/reservation.py (hunk 5 above)
Comment (zaibon) [reply templates/reservation/reservation.py]: don't forget to remove it

**Review by zaibon — CHANGES_REQUESTED** [review]

File: templates/grid_broker/grid_broker.py (hunk 8 above)
Comment (serboctor) [reply templates/grid_broker/grid_broker.py]: First of all, I didn't change the logic at all, I only restructured it so we don't have a refund block twice and hardened it so there is a try/except around the refund. It already did used to refund if there was no data. I don't understand the relationship between data being empty and not being able to refund. This is the data from the notary, the refund function doesn't use the data anyway. Also, if we can attempt to refund when the data can't be parsed, then we can definitely attempt to refund when the data is empty because when parsing fails, we also end up with no data. I just raise a value error so I can take advantage of the refund in the try/except instead of having duplicate code just two lines above it. The `malformed` variable is just there for sake of logging so I can differentiate between both cases.

Comment (zaibon) [reply templates/grid_broker/grid_broker.py]: yeah I got confused, the data is indeed not needed to refund, just to notify the user via email.

**Review by zaibon — APPROVED** [review]
